### PR TITLE
JAMES-3172 Reactify tasks

### PR DESCRIPTION
**Pull request by chibenwa**
Fixes: https://github.com/linagora/james-project/pull/3355

### CODE REVIEW

**Comment by chibenwa** [comment]
> I don't understand how changing execution from Executor to reactor fixed it. Could you explain?

Well, the bet here was that `java8-futures < reactor` for representing asynchronous computation. The diagnostic showed that the `.block()` method was not propagating well to the other reactor scheduler the InteruptedException sent by "future cancellation".

The plan was to:
 - 0. Defines reactive tasks
 - 1. Rely on reactor for task execution
 - 2. So that reactive tasks could be a top level definition
 - 3. When executing such tasks .block is never called and I hoped it would solve the issue.

It turns out that part 1 of the plan was enough to fix the issue. Not sure why.

To be honnest I don't care. I can't understand the full complexity of a beast like reactor, and I am not interested by it.

I prooved the problem, had an hypothesis,  prooved that it fixed the problem I previously demonstrated. The why is not that important, as this piece of software is prooved to be more reliable thus better than before.

File: server/task/task-api/pom.xml [hunk 1]
```diff
@@ -47,6 +47,10 @@
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
```
Comment (mbaechler) [inline server/task/task-api/pom.xml:52]: why did you need that? It looks like a violation of the "no reactor in public APIs" rule

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java [hunk 2]
```diff
@@ -40,22 +34,23 @@
...
     private final Listener listener;
-    private final AtomicReference<Tuple2<TaskId, Future<?>>> runningTask;
+    private final AtomicReference<Tuple2<TaskId, Disposable>> runningTask;
     private final Set<TaskId> cancelledTasks;
     private final Duration pollingInterval;
 
     public SerialTaskManagerWorker(Listener listener, Duration pollingInterval) {
         this.pollingInterval = pollingInterval;
-        this.taskExecutor = Executors.newSingleThreadExecutor(NamedThreadFactory.withName("task executor"));
+        this.taskExecutor = Schedulers.newBoundedElastic(1, 1, "task executor");
```
Comment (mbaechler) [inline server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:53]: why did you choose that specific scheduler and why did you limit the scheduler queue?

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java [hunk 3]
```diff
@@ -107,30 +104,30 @@ public SerialTaskManagerWorker(Listener listener, Duration pollingInterval) {
...
-            .run()
-            .onComplete(result -> Mono.from(listener.completed(taskWithId.getId(), result, taskWithId.getTask().details())).block())
-            .onFailure(() -> {
-                LOGGER.error("Task was partially performed. Check logs for more details. Taskid : " + taskWithId.getId());
-                Mono.from(listener.failed(taskWithId.getId(), taskWithId.getTask().details())).block();
-            });
+    private Mono<Task.Result> runTask(TaskWithId taskWithId, Listener listener) {
+        return Mono.from(Task.toReactiveTask(taskWithId.getTask())
+            .runReactive())
+            .doOnNext(result2 -> result2.onComplete(result -> Mono.from(listener.completed(taskWithId.getId(), result, taskWithId.getTask().details())).block())
```
Comment (mbaechler) [inline server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:118]: result2?

File: server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java [hunk 4]
```diff
@@ -162,17 +161,16 @@ void theWorkerShouldReportThatATaskIsInProgress() throws InterruptedException {
...
         await(taskLaunched);
         verify(listener, atLeastOnce()).started(id);
         verifyNoMoreInteractions(listener);
         latch.countDown();
     }
 
-    @Disabled("JAMES-3172 We cannot cancel computation started by Reactor")
     @Test
-    void reactiveTasksShouldBeCancellable() throws InterruptedException {
+    void tasksCallingReactorShouldBeCancellable() throws InterruptedException {
```
Comment (mbaechler) [inline server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java:173]: what does this test name means?

File: server/task/task-api/pom.xml (hunk 1 above)
Comment (chibenwa) [reply server/task/task-api/pom.xml:52]: (I will drop that commit in a separate PR to not delay the fix)

Well because I follow the same pattern for reactive adoption than what we did for mailbox reactive mailbox listener.

What I believed was the 'state of the art' for progressive reactive code adoption.

This is needed:
 - So that a ReactiveTask can provide a default event method (blocker)
 - So that we can provide a TaskReactiveWrapper (not a blocker: can be moved to the executor)

>  It looks like a violation of the "no reactor in public APIs" rule

Given the above content maybe the rule should be:

 - No reactor Mono / Flux in public API interface signature
 - But it's use is ok in default methods, so we have the right to have it as part of dependencies of API maven modules

WDYT?

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java (hunk 2 above)
Comment (chibenwa) [reply server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:53]: I wanted to use `single` but it rejected calls to `.block` which is a no-go here.

A single thread bounded ealstic scheduler have no such limitations.

File: server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java (hunk 4 above)
Comment (chibenwa) [reply server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java:173]: A task calling reactor should be cancellable

Maybe `taskCallingReactorShouldDisposeReactorResourcesUponCancel` ?

File: server/task/task-api/pom.xml (hunk 1 above)
Comment (mbaechler) [reply server/task/task-api/pom.xml:52]: > > It looks like a violation of the "no reactor in public APIs" rule
> 
> Given the above content maybe the rule should be:
> 
>     * No reactor Mono / Flux in public API interface signature
> 
>     * But it's use is ok in default methods, so we have the right to have it as part of dependencies of API maven modules
> 
> 
> WDYT?

Doing that always ends the same way: we put reactor types everywhere by mistake.

I propose to write a wrapping method and duplicates in every maven module that needs it: there's almost not logic in it so the risk of bugs it very thin, it's tested as part of higher level testsuite and it will disappear as we do the migration.

The cost/risk of the duplication is in my opinion less than the cost/risk of introducing a depency on a library we don't want to depend on

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java (hunk 2 above)
Comment (mbaechler) [reply server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:53]: you answered only half of the question: why a limit of the queue? BTW, you can create a Scheduler from an Executor if needed.

File: server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java (hunk 4 above)
Comment (mbaechler) [reply server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java:173]: the problem is "call reactor", it's not clear to me what it means

File: server/task/task-api/pom.xml (hunk 1 above)
Comment (chibenwa) [reply server/task/task-api/pom.xml:52]: > Doing that always ends the same way: we put reactor types everywhere by mistake.

You being so pessimistic about a rule as simple as `don't use reactor in API interfaces` that I believe everyone of us could easily enforce now that it is clearly established depresses me.

I don't see the added value of the maven dependency enforcement: it is as easy in my opinion to enforce `No reactor Mono / Flux in public API interface signature`

Now what's the associated risk? Turning Mono/Flux into Publisher in case of abuse is no hard work and a pretty easy refactoring. I have no doubt that we could easily cary it out.

Now when I compare the benefits (not writing run() for every task) to the exposure (low) and the consequences (limited) I don't see why I should refrain.

> I propose to write a wrapping method and duplicates in every maven module that needs it: there's almost not logic in it so the risk of bugs it very thin, it's tested as part of higher level testsuite and it will disappear as we do the migration.

Well if we are down to that I still prefer wrinting the .block call in every class.

> The cost/risk of the duplication is in my opinion less than the cost/risk of introducing a depency on a library we don't want to depend on

I see that we have a diverging analyses here.

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java (hunk 2 above)
Comment (chibenwa) [reply server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:53]: > BTW, you can create a Scheduler from an Executor if needed.

Let's do it then

File: server/task/task-api/pom.xml (hunk 1 above)
Comment (mbaechler) [reply server/task/task-api/pom.xml:52]: > I see that we have a diverging analyses here.

Because I did several cleanups in the past because of this accidental dependency and it was not always simple tasks.

It's not always as easy as changing a return type and it costed us much time in the past,

Even cases where it was simple, it was tedious.

Comment (chibenwa) [reply server/task/task-api/pom.xml:52]: > Because I did several cleanups in the past because of this accidental dependency and it was not always simple tasks.

How established was the practice? How much do people care.

Have mentalities and awareness evolved since then?

Comment (chibenwa) [reply server/task/task-api/pom.xml:52]: > Well if we are down to that I still prefer wrinting the .block call in every class.

I will do that tomorrow even if that outcome don't please me at all.

Comment (mbaechler) [reply server/task/task-api/pom.xml:52]: > > Because I did several cleanups in the past because of this accidental dependency and it was not always simple tasks.
> 
> How established was the practice? How much do people care.
> 
> Have mentalities and awareness evolved since then?

I can't really tell, sorry

File: server/protocols/webadmin/webadmin-jmap/src/main/java/org/apache/james/webadmin/data/jmap/RecomputeUserFastViewProjectionItemsTask.java [hunk 5]
```diff
@@ -118,15 +119,15 @@ public String getUsername() {
...
+        return corrector.correctUsersProjectionItems(progress, username)
             .subscribeOn(Schedulers.elastic())
-            .block();
-
-        if (progress.failed()) {
-            return Result.PARTIAL;
-        }
-        return Result.COMPLETED;
+            .map(any -> {
+                if (progress.failed()) {
```
Comment (blackheaven) [inline server/protocols/webadmin/webadmin-jmap/src/main/java/org/apache/james/webadmin/data/jmap/RecomputeUserFastViewProjectionItemsTask.java:126]: I do not get what happen here, is it blocking?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-jmap/src/main/java/org/apache/james/webadmin/data/jmap/RecomputeUserFastViewProjectionItemsTask.java:126]: This code relies on a side effect rather than on a map + reduce.

I don't intend to change that here...

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-jmap/src/main/java/org/apache/james/webadmin/data/jmap/RecomputeUserFastViewProjectionItemsTask.java:126]: This is fixed here: https://github.com/linagora/james-project/pull/3377/commits/1ee8a35da336564efbf1aa15062d75ff146b7fde

**Comment by chibenwa** [comment]
I don't think this makes much sense, and I don't want to invest time to finish to contribute this.
